trinity/sync/full/trie.py

```diff
@@ -0,0 +1,231 @@
+        else:
+            self._schedule(root_hash, parent=None, depth=0, leaf_callback=self.leaf_callback)
+
+    async def leaf_callback(self, data: bytes, parent: SyncRequest) -> None:
+        """Called when we reach a leaf node.
+
+        Should be implemented by subclasses that need to perform special handling of leaves.
+        """
+        pass
+
+    @property
+    def has_pending_requests(self) -> bool:
+        return len(self.requests) > 0
+
+    def next_batch(self, n: int = 1) -> List[SyncRequest]:
+        """Return the next requests that should be dispatched."""
+        if len(self.queue) == 0:
+            return []
+        batch = list(reversed((self.queue[-n:])))
+        self.queue = self.queue[:-n]
+        return batch
+
+    async def schedule(self, node_key: Hash32, parent: SyncRequest, depth: int,
+                       leaf_callback: Callable[[bytes, 'SyncRequest'], Awaitable[None]],
+                       is_raw: bool = False) -> None:
+        """Schedule a request for the node with the given key."""
+        if node_key in self._existing_nodes:
+            self.logger.debug("Node %s already exists in db" % encode_hex(node_key))
+            return
+        if await self.db.coro_exists(node_key):
+            self._existing_nodes.add(node_key)
+            self.logger.debug("Node %s already exists in db" % encode_hex(node_key))
+            return
+        self._schedule(node_key, parent, depth, leaf_callback, is_raw)
+
+    def _schedule(self, node_key: Hash32, parent: SyncRequest, depth: int,
+                  leaf_callback: Callable[[bytes, 'SyncRequest'], Awaitable[None]],
+                  is_raw: bool = False) -> None:
+        if parent is not None:
+            parent.dependencies += 1
+
+        existing = self.requests.get(node_key)
+        if existing is not None:
+            self.logger.debug(
+                "Already requesting %s, will just update parents list" % node_key)
+            existing.parents.append(parent)
+            return
+
+        request = SyncRequest(node_key, parent, depth, leaf_callback, is_raw)
+        # Requests get added to both self.queue and self.requests; the former is used to keep
+        # track which requests should be sent next, and the latter is used to avoid scheduling a
+        # request for a given node multiple times.
+        self.logger.debug("Scheduling retrieval of %s" % encode_hex(request.node_key))
+        self.requests[request.node_key] = request
+        bisect.insort(self.queue, request)
+
+    def get_children(self, request: SyncRequest
+                     ) -> Tuple[List[Tuple[int, Hash32]], List[bytes]]:
+        """Return all children of the node retrieved by the given request.
+
+        :rtype: A two-tuple with one list containing the children that reference other nodes and
+        another containing the leaf children.
+        """
+        node = decode_node(request.data)
```

Comment:
This stands out as a potential source of trouble.  `request.data` can be `None` in cases where the `SyncRequest` has not been processed.  I trust that this code path is only called *after* the `data` attribute has been populated, but the implicitness of this approach makes me think we should look for an alternate pattern, maybe something like a `SyncResponse` object which takes the request and the data as an argument and exposes `response.data` which is always guaranteed to be populated.

Reply:
get_children() is called precisely to find the children of the node we're processing, and `process()` does set `request.data` immediately before calling it. I guess if I inlined this method into `process()` it wouldn't raise a red flag, and we wouldn't make `process()` any more complex or less readable as this is a two-line method that just decodes the request data and passes it to `_get_children()`

Reply:
:+1: if it gets rid of the mutation of the `request` object and thus the implicit availability of the `request.data` attribute.

Reply:
In order to completely get rid of the mutation of `request.data` I'll have to get rid of the check in `process()` that raises `SyncRequestAlreadyProcessed` when request.data is not None, but that is useless anyway as it is caught and ignored in `StateDownloader`

Reply:
Hmm, spoke too soon. I thought I'd be able to have `process()` pass the `data` to `commit()` but that doesn't work because we don't always commit at the end of `process()` -- we only commit once all the child nodes have been retrieved and committed. IOW, we need to keep (somewhere) the data received for every request, until that request is eventually committed. I could get rid of `request.data`, but then I'd have to store the data for every request in an instance variable of `HexaryTrieSync`. The `SyncResponse` you suggest would also have to be stored somewhere, and since it's disjoint from the actual request I'm not sure it gives us any guarantees really -- we could still try to commit a request for which there's no `SyncResponse` in the same way we can try to commit one which has `.data` set to None

Reply:
If you can't find a way to do it that seems good, my fallback request would be to add some comments to document this deficiency.  Given that if `request.data` is `None` the two methods that access it will very likely fail loudly I'm ok with it staying in.

Reply:
Ok, I've added a docstring to `commit()`, and the other method is gone as there was no point in having it anyway